### PR TITLE
Added protected peers functionality

### LINKED ISSUE #1275: implement protected peers functionality
## Task summary

- there are peers who we wish to maintain connections to (ie. don't prune them when our peer count is over maximum)
- the ConnManager should be updated as to not disconnect from these peers (see `ConnManager.Connected`)
- a peer should be added to this list when we are syncing with them (they can be removed when syncing with them is completed)

## Specification reference

- n/a

## Other information and links

- I believe substrate has this functionality implemented, can look there for some reference

### PR DESCRIPTION
**Pull request by anthdm**
Added protected peers functionality and prevent syncing peers from pruning when the maximum amount of connecting peers is exceeded.

## Changes

- Updated `ConnManager` so it can protect and unprotect peers
- Updated `ConnManager` so it will not prune protected peers
- Updated `Service` so it will use the `ConnManager` to flag and unflag syncing peers.

## Tests

```

```

## Checklist

- [x] I have read [CODE_OF_CONDUCT](https://github.com/ChainSafe/gossamer/blob/development/.github/CODE_OF_CONDUCT.md) and [CONTRIBUTING](https://github.com/ChainSafe/gossamer/blob/development/.github/CONTRIBUTING.md) 
- [x] I have provided as much information as possible and necessary
- [x] I have reviewed my own pull request before requesting a review
- [ ] All integration tests and required coverage checks are passing

## TODO
- Add test that checks if peers are flagged protected from the `Service`

## Issues

- closes #1275

### CODE REVIEW

**Review by RyRy79261 — APPROVED** [review]

**Review by arijitAD — COMMENTED** [review]
.

File: dot/network/connmgr.go [hunk 1]
```diff
@@ -73,18 +79,40 @@ func (*ConnManager) GetTagInfo(peer.ID) *connmgr.TagInfo { return &connmgr.TagIn
...
 
-// Unprotect peer
-func (*ConnManager) Unprotect(peer.ID, string) bool { return false }
+	cm.protectedPeerMap[id] = struct{}{}
+}
+
+// Unprotect peer will remove the given peer from prune protection.
+// returns true if we have successfully removed the peer from the
+// protectedPeerMap. False otherwise.
+func (cm *ConnManager) Unprotect(id peer.ID, tag string) bool {
```
Comment (arijitAD) [inline dot/network/connmgr.go:94]: Would it be good to rename `Unprotect ` to `UnprotectPeer` for more clarity?

Comment (arijitAD) [reply dot/network/connmgr.go:94]: Same for `Protect`

File: dot/network/connmgr.go [hunk 2]
```diff
@@ -73,18 +79,40 @@ func (*ConnManager) GetTagInfo(peer.ID) *connmgr.TagInfo { return &connmgr.TagIn
...
+	cm.protectedPeerMapMu.Lock()
+	defer cm.protectedPeerMapMu.Unlock()
+
+	_, ok := cm.protectedPeerMap[id]
+	if ok {
+		delete(cm.protectedPeerMap, id)
```
Comment (arijitAD) [inline dot/network/connmgr.go:100]: Can you remove the return type? So we can directly call delete without checking for key first.

**Review by arijitAD — COMMENTED** [review]
Having multiple locks in ConnManager would introduce additional complexity. Can you use the existing mutex from ConnManager?

File: dot/network/connmgr.go (hunk 1 above)
Comment (noot) [reply dot/network/connmgr.go:94]: the functions have to be named what they are to satisfy the libp2p connection manager interface I believe

**Review by noot — COMMENTED** [review]
great work!

could you also add the peer protection in `beginSyncing` and `handleBlockAnnounceMessage`? it would be nice to have both the `s.syncing` map set and `ConnMananger.Protect` in one function since they both happen together

 it would be nice to add a test that runs 2 network.Services and sends a BlockRequest message to the other via `beginSyncing` and makes sure they are protected

File: dot/network/connmgr.go [hunk 3]
```diff
@@ -115,14 +154,18 @@ func (cm *ConnManager) Connected(n network.Network, c network.Conn) {
...
-		i := rand.Intn(len(n.Peers())) //nolint
-		peers := n.Peers()
-		logger.Trace("Over max peer count, disconnecting from random peer", "peer", peers[i])
-		err := n.ClosePeer(peers[i])
+		var (
+			unprotPeers = cm.unprotectedPeers(n.Peers())
+			// TODO: change to crypto/rand
+			i = rand.Intn(len(unprotPeers)) //nolint
+		)
+
```
[the remark this answers is not in the document]
Comment (anthdm) [reply dot/network/connmgr.go:164]: good catch

**Review by anthdm — COMMENTED** [review]
@arijitAD If we use the existing mutex we will have a never-ending lock. I tested this before. Using a new RWMutex will let us lock and unlock it only for reading in most cases also.

**Comment by anthdm** [comment]
> handleBlockAnnounceMessage

Good idea! Will update the PR with those suggestions.

**Comment by anthdm** [comment]
> great work!
> 
> could you also add the peer protection in `beginSyncing` and `handleBlockAnnounceMessage`? it would be nice to have both the `s.syncing` map set and `ConnMananger.Protect` in one function since they both happen together
> 
> it would be nice to add a test that runs 2 network.Services and sends a BlockRequest message to the other via `beginSyncing` and makes sure they are protected

Good idea! I updated the PR with those suggestions.

**Review by noot — APPROVED** [review]
great work!

**Pull request merged**